OGraf Server API

nytamin opened the pull request:
This PR tracks the work being done to define the **Server API portion of the OGraf specification**.

## Time plan

_(Revised september 2025)_

* **Summer 2025**: Work on an initial draft within the [HTML Graphics Working Group](https://tech.ebu.ch/groups/html_graphics)
* **Autumn 2025**: First public draft announced & feedback from the industry / vendors.
* **Q1 2026**: Finalize first version of the Server API (i.e. merge this PR)

## Info

The Server API is an OpenApi definition, located at [v1/specification/open-api/server-api.yaml](https://github.com/ebu/ograf/blob/feat/server-api/v1/specification/open-api/server-api.yaml)
(also see [the auto-generated documentation here](https://html-preview.github.io/?url=https://github.com/ebu/ograf/blob/feat/server-api/v1/specification/open-api/docs/index.html)).

## Contributions & testing

Ways to contribute:

* Join the discussions in the [HTML Graphics Working Group](https://tech.ebu.ch/groups/html_graphics)
* Comments on this PR thread will be forwarded to the working group.

I (@nytamin ) intend to update SuperFly's ["Simple Rendering System"](https://github.com/SuperFlyTV/ograf-server) continously throughout the discussions and keep it up-to-date with the latest server API, for testing and validation purposes.

nytamin committed:
feat: OpenApi definition for Server API

nytamin committed:
fix: serverApi spec: delete cannot have requestBody

nytamin committed:
chore: wip

nytamin committed:
fix: changes to serverApi after group talk yesterday

nytamin committed:
fix: serverApi: add target info endpoint

nytamin committed:
fix: change a get to a post

not emantically correct, but technically possible

nytamin committed:
fix: server-API: change a few posts to puts

nytamin committed:
fix: serverApi: move renderTarget to queryParameters

nytamin committed:
chore: rename to v1

nytamin committed:
Merge branch 'main' into feat/server-api

# Conflicts:
#	v1-draft-0/examples/l3rd-name/manifest.json
#	v1/examples/l3rd-name/l3rd-manifest.json
#	v1/examples/l3rd-name/manifest.json
#	v1/specification/docs/Specification.md
#	v1/typescript-definitions/package-lock.json
#	v1/typescript-definitions/package.json

nytamin committed:
chore: update openapi docs

nytamin committed:
Merge branch 'main' into feat/server-api

# Conflicts:
#	v1/typescript-definitions/src/apis/graphicsAPI.ts

nytamin committed:
fix: remove the option to address a graphic by its id, in favor for a GraphicInstanceId only

nytamin committed:
Merge branch 'main' into feat/server-api

nytamin committed:
chore: update generated types

nytamin committed:
Merge branch 'main' into feat/server-api

nytamin commented:
Notes from Working Group meeting 2025-10-01:

There was a discussion regarding PR #15 and whether we should simplify the API to only allow 1 graphic per RenderTarget.
There are both pros and cons with allowing only 1 Graphic per RenderTarget:
* pro: This simplifies the API, since you wouldn't have to address a Graphic individually so send commands to it, only the RenderTarget (like how CasparCG works, adressing commands to a "layer").
* con: It adds restrictions on how a Renderer must operate.
  _It was noted that Ross has a way of transitioning (dissolve) between gfx, which would imply temporarily having multiple Graphics at the same time._
* con: If we have GraphicInstanceIds, it allows controllers to use a Render "without RenderTargets", thus only adressing the Graphics themselves.

No conclusion was reached, we'll discuss this again at next meeting in two weeks.

dvdeurse commented:
I went through the OpenAPI definition and had the following remarks:
* general
  * the Error response schema now consists of 2 fields (`error` and `stack`). Shouldn't we look into a more standardized way for representing errors (e.g. https://www.rfc-editor.org/rfc/rfc7807.html)?
  * date(time) fields are currently represented by a Unix timestamp. I believe UTC-based ISO 8601 strings are more common in today's public JSON-based HTTP APIs. They are much more readable and there is plenty of support in the libraries and frameworks.
* Server information endpoint (`GET /`)
  * I would omit `uptime`, it has nothing to do with OGraf 
* Graphic Instance model
  * there is now a distinction between a Graphic and a Manifest, where the Graphic is a subset of the information available in the Manifest (except for the `created` and `modified` fields). Wouldn't it be cleaner to just have a full version of the Graphic (which corresponds to the content of the Manifest) and a light version (which is a subset of the Manifest) to be used in the list endpoint?
  * I'm not sure about the `created` and `modified` fields. I understand this kind of information is meaningful for clients, but it limits the way you can represent the provenance information, mostly because of the existing `author` field. There is now no way to distinguish between the original creator and someone else who did an update. Maybe we need a field for the `graphic` (which is exactly the field of the previous point) and a field for the provenance which then would contain the fields `createdAt`, `createdBy`, `updatedAt`, `updatedBy`. This way, the `author` field in the manifest can refer to an organization or an individual, while the provenance fields can refer to individual users.
* Renderer (`GET /renderers/{rendererId}`)
  * shouldn't we split the `status` field into `status` and `renderTargets`? 
* Renderer target (`GET /renderers/{rendererId}/target`)
  * why is there a surrounding `renderTarget` element in the response?
  * I'm not sure a target needs a `status` and `statusMessage` ... Any valid use cases you can think of?
* Graphic actions (play/update/stop/custom/gotoTime/setActionSchedule)
  * depending on the GraphicInstanceId vs GraphicId discussion for the identification inside a target, but if we choose the GraphicInstanceId path, I wonder whether we still need the renderTarget to address a graphic instance (given we already have the GraphicInstanceId). Of course, this would mean that the GraphicInstanceId is unique across a Renderer, not only across a target.
  * I would also add the Graphic identification (whether it will be GraphicsInstanceId, RendererTarget, GraphicsId or a combination) into the body, instead of using query params
  * if we would go for GraphicsInstanceId identification only, I would also propose to change the URI to `/renderers/{rendererId}/graphic` (so omitting the 'target' part), as in this case, you execute an action on a Graphic, not on a target. This approach would allow you to explain both Renderers that don't care about targets and Renderers that do.

nytamin committed:
fix(serverApi): add skipAnimation to match graphicsApi

nytamin commented:
Notes from Working Group meeting 2025-10-29:

* Discussion on whether to have one or multiple graphics per renderTarget. We concluded with that since we have no definite arguments for or against, we'll go with the generic approach for now (ie multiple graphics per RenderTarget).
* It might be useful if a Renderer can specify whether it renders one or multiple graphics per renderTarget.
* Davy's proposals look good!

nytamin committed:
Merge branch 'main' into feat/server-api

# Conflicts:
#	v1/typescript-definitions/package-lock.json
#	v1/typescript-definitions/package.json
#	v1/typescript-definitions/src/main.ts

nytamin committed:
chore: ci

nytamin committed:
chore: rename server-api.d.ts -> server-api.ts so that it is included in built

nytamin committed:
Merge branch 'main' into feat/server-api

# Conflicts:
#	.github/workflows/publish-npm.yml

nytamin committed:
fix: make graphicInstances required

nytamin committed:
Merge pull request #15 from ebu/fix/graphicInstanceId

ServerAPI: Remove GraphicId in favor for GraphicInstanceId

nytamin committed:
fix: make properties of RenderTargetInfo be required

nytamin committed:
chore: build artifacts

nytamin committed:
fix: change errorResponses to align with RFC7807: Problem Details for HTTP APIs

ref: https://www.rfc-editor.org/rfc/rfc7807.html

nytamin committed:
fix: replace author, created and modified fields with createdAt, createdBy, updatedAt, updatedBy fields

nytamin committed:
fix: remove uptime

nytamin committed:
fix: move renderTargets out of status property

nytamin committed:
fix: move renderTargetInfo out to parent object

nytamin committed:
fix: remove renderTarget status and statusMessage

( we can re-add them later if we come up with a use case)

nytamin committed:
fix: rename graphicTarget to graphicInstanceId

nytamin committed:
fix: remove `graphic` from ClearGraphicsResponse

nytamin committed:
chore: update artifacts

nytamin commented:
I pushed a few modifications to this PR based on @dvdeurse 's suggestions.

Notable exceptions:

> Wouldn't it be cleaner to just have a full version of the Graphic (which corresponds to the content of the Manifest) and a light version (which is a subset of the Manifest) to be used in the list endpoint?

I couldn't come up with a good solution for this. Let's discuss it some time.

> I wonder whether we still need the renderTarget to address a graphic instance (given we already have the GraphicInstanceId). Of course, this would mean that the GraphicInstanceId is unique across a Renderer, not only across a target.

I think it would be useful to keep the graphicsInstanceId being unique only within its RenderTarget. This way we allow for a Renderer that _doesn't_ really support instances of a graphic to function, by just returning the `graphicId` as graphicInstanceId upon load.

dvdeurse commented:
Thx Johan for these modifications. Some remarks:

>> Wouldn't it be cleaner to just have a full version of the Graphic (which corresponds to the content of the Manifest) and a light version (which is a subset of the Manifest) to be used in the list endpoint?

> I couldn't come up with a good solution for this. Let's discuss it some time.

In the end, for a single Graphic, there are two parts:
* information from the manifest
* information from the server, for example when a Graphic was uploaded (createdAt), by whom (createdBy) etc. Typically provenance information, but maybe there are other use cases. The idea is that you can upload a Graphic to multiple servers. The manifest part will always be the same, but the additional server-specific part will differ. One approach could also be that we omit this part from the spec and leave that up to the vendors.

Then there's the info from the manifest itself. Do we want a lightweight form of it? I believe we do, for two reasons:
* keep the list endpoint light
* when referencing a Graphic (e.g. in the list of Graphic Instances loaded in a target), you don't want to provide the full manifest

Therefore, I propose the following data model for a Graphic, where `server` is to be discussed whether we want this in the spec or not (and if yes, the name of the field `server` might be improved as well):
```
{
  "graphic": {
    "$schema": "string",
    "id": "string",
    "version": "string",
    "main": "string",
    "name": "string",
    "description": "string",
    "author": {
      "name": "string",
      "email": "string",
      "url": "string"
    },
    "customActions": [
      "string"
    ],
    "supportsRealTime": true,
    "supportsNonRealTime": true,
    "stepCount": 1,
    "schema": {
    },
    "renderRequirements": [
      {
        "resolution": {
          "width": "string",
          "height": "string"
        },
        "frameRate": "string",
        "accessToPublicInternet": "string"
      }
    ]
  },
  "server": {
    "createdAt": "...",
    "createdBy: {
      ...
    }
  }
}
```

That is the full version, the lightweight version of a Graphic would then be:
```
{
  "id": "string",
  "name": "string",
  "description": "string"
}
```

> I think it would be useful to keep the graphicsInstanceId being unique only within its RenderTarget. This way we allow for a Renderer that doesn't really support instances of a graphic to function, by just returning the graphicId as graphicInstanceId upon load.

Agreed.

<br>

On the clearGraphic endpoint, there is an error in the request data model. I believe you intended to implement an array of filters, but the schema currently models `filters` as an object.

<br>

Finally, there was one remark that you forgot to implement or mention:
> I would also add the Graphic identification (whether it will be GraphicsInstanceId, RendererTarget, GraphicsId or a combination) into the body, instead of using query params

This remark was about the query params on the following endpoints:
* clearGraphic
* loadGraphic
* updateAction
* playAction
* customAction
* gotoTime

I don't see a reason why they should be query params instead of being part of the request body. The only place where the query param makes sense is on the `GET /renderers/{rendererId}/target` (as you cannot provide a body in a GET request).

nytamin committed:
Merge branch 'main' into feat/server-api

nytamin committed:
fix: rework what graphics info and manifest data is returned at various endpoints

nytamin committed:
fix: change filters to be an array of objects

nytamin committed:
fix: move `renderTarget` and `graphicInstanceId` from `query` to the `body`, where applicable

nytamin committed:
fix: minor fixes to server-api

nytamin committed:
chore: dist

nytamin commented:
Good suggestions @dvdeurse !

I've updated the schema to what you proposed:

* I chose the name `metadata` instead of `server` for the graphics endpoint. I think it fits better, but I'm open for further suggestions!
* I changed the `filters` on the clearGraphics endpoint, good spot!
* I moved the `renderTarget` and `graphicInstanceId` from `query` to the `body`, where applicable.

nytamin committed:
Merge branch 'main' into feat/server-api

nytamin committed:
fix: changes after review in PR:

* change PUT to POST for `/renderers/{rendererId}/target/graphic/load`
* change `/renderers/{rendererId}/target/graphic/customAction` to `/renderers/{rendererId}/target/graphicInstance/customActions/{customActionId}:`
* change all `/renderers/{rendererId}/target/graphic/*` to `/renderers/{rendererId}/target/graphicInstance/*`

Also, the following change:

* Add `550` response code, to signify the difference of an "internal server error" (500) and an error thrown from a GraphicsInstance (550)

nytamin committed:
chore: dist

nytamin committed:
chore: typo

nytamin committed:
fix: align with graphics api

nytamin commented:
Notes from Working Group meeting 2026-02-04:

**Regarding the suggestion to remov non-realtime API endpoints:**

The group agrees that non-realtime Renderers would probably not be controlled via the Server API but instead via some proprietary control API. It makes sense to remove the non-realtime endpoints from the Server API.

**Regarding releasing the Server API draft:**

The group thinks that the API is mature enough to be released as a draft.

Todo:

- [ ] Update documentation to highlight that the Server API is a draft.
- [ ] Merge this PR. Subsequent changes to be put in separate PRs.

nytamin committed:
fix(server-api): remove non-realtime API endpoints.

After decision by the Working Group last week,

nytamin committed:
chore: docs
